[TSAN] Add __tsan_check_no_mutexes_held helper

kennyyu opened the pull request:
This adds a new helper that can be called from application code to ensure that no mutexes are held on specific code paths. This is useful for multiple scenarios, including ensuring no locks are held:

- at thread exit
- in peformance-critical code
- when a coroutine is suspended (can cause deadlocks)

See this discourse thread for more discussion:
https://discourse.llvm.org/t/add-threadsanitizer-check-to-prevent-coroutine-suspending-while-holding-a-lock-potential-deadlock/74051

kennyyu committed:
[TSAN] Add __tsan_check_no_mutexes_held helper

This adds a new helper that can be called from application code
to ensure that no mutexes are held on specific code paths. This
is useful for multiple scenarios, including ensuring no locks
are held:

- at thread exit
- in peformance-critical code
- when a coroutine is suspended (can cause deadlocks)

See this discourse thread for more discussion:
https://discourse.llvm.org/t/add-threadsanitizer-check-to-prevent-coroutine-suspending-while-holding-a-lock-potential-deadlock/74051

dvyukov commented on compiler-rt/include/sanitizer/tsan_interface.h:
```diff
@@ -126,6 +126,10 @@ void __tsan_mutex_post_signal(void *addr, unsigned flags);
 void __tsan_mutex_pre_divert(void *addr, unsigned flags);
 void __tsan_mutex_post_divert(void *addr, unsigned flags);
 
+// Annotate that no mutexes can be held. If we are holding mutexes, then
```
This is effectively not an annotation, but a check. So I would say something along the following lines:

```
// Check that the current thread does not hold any mutexes,
// report a bug report otherwise.
```

kennyyu replied:
Will change

dvyukov commented on compiler-rt/test/tsan/mutex_cannot_be_locked.cpp:
```diff
@@ -0,0 +1,24 @@
+// RUN: %clangxx_tsan -O1 %s -o %t && %deflake %run %t | FileCheck %s
+#include "test.h"
+
+pthread_mutex_t mtx;
```
It needs to be initialized.

kennyyu replied:
Will fix

dvyukov commented on compiler-rt/test/tsan/mutex_cannot_be_locked_success.cpp:
```diff
@@ -0,0 +1,19 @@
+
+int main() {
+  pthread_t th;
+  pthread_create(&th, 0, ThreadFunc, NULL);
```
Creating threads does not look necessary in these tests.

kennyyu replied:
Will remove

dvyukov commented on compiler-rt/lib/tsan/rtl/tsan_report.cpp:
```diff
@@ -216,11 +218,16 @@ static void PrintMutexShortWithAddress(const ReportMutex *rm,
+    Printf("  Mutex M%u (%p) created at:\n", rm->id,
+           reinterpret_cast<void *>(rm->addr));
+  } else {
+    Printf("  Mutex M%u (%p) acquired at:\n", rm->id,
+           reinterpret_cast<void *>(rm->addr));
+  }
```
I think it still will be creation stack (in the test you got lock stack because you forgot to init the mutex). Or it may be a lock stack in a different thread/context.
It's basically the stack that somehow identifies the mutex for the user.
So I don't think this part of the change is necessary.

kennyyu replied:
Will revert

dvyukov commented on compiler-rt/lib/tsan/rtl/tsan_report.cpp:
```diff
@@ -93,7 +93,9 @@ static const char *ReportTypeString(ReportType typ, uptr tag) {
       return "lock-order-inversion (potential deadlock)";
-    // No default case so compiler warns us if we miss one
+    case ReportTypeMutexCannotBeLocked:
+      return "mutex cannot be locked on this code path";
```
The word "locked" makes me think that the report is produces when a mutex is locked. We are not locking a mutex here, but rather "holding" it. The word "held" used in the interface function name looks like a better fit here. Perhaps something like "mutex held in the wrong context"?

kennyyu replied:
Will update wording

dvyukov reviewed (commented):
Some cosmetic comments, but overall looks good.

kennyyu committed:
fixup! [TSAN] Add __tsan_check_no_mutexes_held helper

- "locked" -> "held". Rename functions, enum value, and test cases to match
- init mutex and don't use threads in tests
- undo mutex acquired/created change
- fix comment on interface function

kennyyu reviewed (commented):
Thanks for the suggestions @dvyukov !

kennyyu commented:
@zmodem Thanks for catching it! Sorry for the breakage! This is my first time contributing to llvm, and I wasn't sure which signals to verify. I'll resubmit the PR with the fix.

kennyyu commented:
https://github.com/llvm/llvm-project/pull/71568 resubmits this PR with the fix.
